lib/theme_check/language_server/completion_providers/filter_completion_provider.rb

```diff
@@ -26,32 +30,63 @@ def can_complete?(content, cursor)
 
       private
 
-      def available_labels
-        @labels ||= ShopifyLiquid::Filter.labels - ShopifyLiquid::DeprecatedFilter.labels
+      def context_with_cursor_before_potential_filter_separator(context)
+        content = context.content
+        diff = content.index(FILTER_SEPARATOR_INCLUDING_SPACES) - content.size
+
+        return context unless content.scan(FILTER_SEPARATOR_INCLUDING_SPACES).size == 1
+
+        context.clone_and_overwrite(col: context.col + diff)
+      end
+
+      def determine_input_type(context)
+        variable_lookup = VariableLookupFinder.lookup(context)
+
+        if variable_lookup
+          object, property = VariableLookupTraverser.lookup_object_and_property(variable_lookup)
+          return property.return_type if property
+          return object.name if object
+        end
+      end
+
+      def available_labels_for(input_type)
+        filters = ShopifyLiquid::SourceIndex.filters
+          .filter_map do |filter|
+          filter.name if input_type.nil? ||
+            filter.input_type == input_type
+        end
+        return available_labels_for(nil) if filters.empty?
+        return filters if input_type == INPUT_TYPE_VARIABLE
+
+        filters.sort + available_labels_for(INPUT_TYPE_VARIABLE).sort
```

Comment:
/nit if im understanding correctly, wouldn't `available_labels_for(INPUT_TYPE_VARIABLE).sort` and `available_labels_for(nil)` return the same thing every time since `nil` and `INPUT_TYPE_VARIABLE` are static? 

if so, are we wasting cycles by computing it every time vs computing this once in a class instance method?

Reply:
Yup, that's correct, thanks for the heads-up! 😬 
I changed the logic a bit, so now it's even more readable :D

Reply:
(documenting here for posterity) after suggesting this i realized using a class instance variable means introducing state, which means if refresh happens and filters change.... the class instance variable will never be updated 😬 so! some extra cycles on filters a small data set is better than managing more state in other parts of code completion code